glusterd2/commands/volumes/brick-replace.go

```diff
@@ -0,0 +1,221 @@
+				subVolIndex = index
+				brickIndex = i
+				srcBrickInfo = brick
+				break LOOP
+			}
+		}
+	}
+
+	excludeZones := make([]string, 0)
+	for svIndex, sv := range subVols {
+		// if SubvolZonesOverlap is true then bricks of only that particular
+		// subvolume will be considered.
+		if req.SubvolZonesOverlap && subVolIndex != svIndex {
+			continue
+		}
+		for _, b := range sv.Bricks {
+			p, err := peer.GetPeer(b.PeerID.String())
+			if err != nil {
+				restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+				return
+			}
+			excludeZones = append(excludeZones, p.Metadata["_zone"])
+		}
+	}
+	req.ExcludeZones = append(req.ExcludeZones, excludeZones...)
+
+	subvolumes := make([]api.SubvolReq, 0)
+	volreq := api.VolCreateReq{
+		Subvols:      subvolumes,
+		Size:         vol.Capacity,
+		LimitPeers:   req.LimitPeers,
+		LimitZones:   req.LimitZones,
+		ExcludePeers: req.ExcludePeers,
+		ExcludeZones: req.ExcludeZones,
+	}
+	availableVgs, err := bricksplanner.GetAvailableVgs(&volreq)
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+	// TODO: check for available vgs in zones already being used in volume.
+	if len(availableVgs) == 0 {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, "No volume groups are available")
+		return
+	}
+
+	mtabEntries, err := volume.GetMounts()
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	// Get source brick information like size etc
+	brickInfo, err := volume.BrickStatus(srcBrickInfo, mtabEntries)
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	// Get new brick from the available vgs
+	newBrick := bricksplanner.GetNewBrick(availableVgs, brickInfo, vol, subVolIndex, brickIndex)
+
+	peerID := uuid.Parse(newBrick.PeerID)
+	if peerID == nil {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, "peer id of new brick could not be parsed")
+		return
+	}
+	allPeerIDs := vol.Nodes()
+	nodes := []uuid.UUID{peerID}
+	txn, err := transaction.NewTxnWithLocks(ctx, volname)
+	if err != nil {
+		status, err := restutils.ErrToStatusCode(err)
+		restutils.SendHTTPError(ctx, w, status, err)
+		return
+	}
+	defer txn.Done()
+
+	txn.Steps = []*transaction.Step{
+		{
+			DoFunc: "brick-replace.PrepareBricks",
+			Nodes:  nodes,
+		},
+		{
+			DoFunc: "brick-replace.ReplaceVolinfo",
+			Nodes:  []uuid.UUID{gdctx.MyUUID},
+		},
+		{
+			DoFunc:   "vol-create.InitBricks",
+			UndoFunc: "vol-create.UndoInitBricks",
+			Nodes:    nodes,
+		},
+		{
+			DoFunc: "vol-start.StartBricks",
+			Nodes:  nodes,
+		},
+		{
+			DoFunc:   "vol-create.StoreVolume",
+			UndoFunc: "vol-create.UndoStoreVolume",
+			Nodes:    []uuid.UUID{gdctx.MyUUID},
+		},
+		{
+			DoFunc: "vol-expand.NotifyClients",
+			Nodes:  allPeerIDs,
+		},
+	}
+
+	if err := txn.Ctx.Set("newBrick", &newBrick); err != nil {
```

Comment:
reuse err variable instead of creating new one always

Reply:
i will make the change, but I fail to see the difference. Does it cause higher memory consumption?

Reply:
currently, the scope of the variable will be limited it creates and destroys the variable each time. its good to use a single error variable.